src/lib/permutations/keccak_perm/keccak_perm.h

```diff
@@ -0,0 +1,125 @@
+* in terms of the Keccak[c] – a detail which cannot be found in [1].
+*
+*
+*
+* [1] FIPS PUB 202 – FEDERAL INFORMATION PROCESSING STANDARDS PUBLICATION – SHA-3 Standard: Permutation-Based Hash and Extendable-Output Functions
+*       https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.202.pdf#page=28
+* [2] https://csrc.nist.gov/projects/hash-functions/sha-3-project
+*/
+
+class Keccak_Permutation final {
+   public:
+       /**
+        * @brief Instantiate a Keccak permutation
+        *
+        * @param capacity_bits Keccak capacity
+        * @param custom_padd the custom bit padding that is to be appended on the call to finish
+        * @param custom_padd_bit_len the bit length of the custom_padd
+        */
+      explicit Keccak_Permutation(size_t capacity_bits, uint64_t custom_padd, uint8_t custom_padd_bit_len);
+
+      size_t hash_block_size() const { return m_bitrate / 8; }
+
+      size_t capacity() const { return m_capacity; }
+
+      size_t bit_rate() const { return m_bitrate; }
+
+      void clear();
+      std::string provider() const;
+
+      secure_vector<uint64_t>& internal_state() { return m_S; }
+
+      const secure_vector<uint64_t>& internal_state() const { return m_S; }
+
+      void set_internal_pos(uint32_t pos) { m_S_pos = pos; }
+
+      uint32_t internal_pos() const { return m_S_pos; }
+
+      // Static functions for internal usage
+
+      /**
+      * Absorb data into the provided state
+      * @param bitrate the bitrate to absorb into the sponge
+      * @param S the sponge state
+      * @param S_pos where to begin absorbing into S
+      * @param input the input data
+      */
+      static uint32_t absorb(size_t bitrate, secure_vector<uint64_t>& S, size_t S_pos, std::span<const uint8_t> input);
+
+      void absorb(std::span<const uint8_t> input);
+
+      /**
+      * Add final padding and permute. The padding is assumed to be
+      * init_pad || 00... || fini_pad
+      *
+      * @param bitrate the bitrate to absorb into the sponge
+      * @param S the sponge state
+      * @param S_pos where to begin absorbing into S
+      * @param custom_padd the custom padding bits used by the primitive derived from Keccak_Permutation
+      * @param custom_padd_bit_len the bit length of the custom padding
+      */
+
+      static void finish(
+         size_t bitrate, secure_vector<uint64_t>& S, size_t S_pos, uint64_t custom_padd, uint8_t custom_padd_bit_len);
+
+      void finish();
+
+      /**
+      * Expand from provided state
+      * @param bitrate sponge parameter
+      * @param S the state
+      * @param output the designated output memory
+      */
+      static void expand(size_t bitrate, secure_vector<uint64_t>& S, std::span<uint8_t> output);
+
+      void expand(std::span<uint8_t> output);
+
+      /**
+      * The bare Keccak[c] permutation
+      */
+      static void permute(uint64_t A[25]);
+      void permute();
```

Comment:
Assuming that `absorb()` and `expand()` can deal with multiple invocations and unaligned input/output data: Is there actually a use case to call `permute()` from using classes? To my mind, this seems to be an implementation detail of the sponge construction.

Reply:
`::permute()` is made private in #3675.